### DISCUSSION TITLE
How to update casks in taps maintained by myself, via `brew upgrade`

### DISCUSSION THREAD

**Discussion by kjxbyz** [Homebrew/discussions#5419] [Tap maintenance and Homebrew development]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.2-24-g443e4e5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 443e4e5dacd047ca86155e252ff7b9c43b01459b
Last commit: 21 hours ago
Core tap JSON: 30 May 12:35 UTC
Core cask tap HEAD: a6d3a98698198edfdc6f02f0cb591822bf228c05
Core cask tap last commit: 3 days ago
Core cask tap JSON: 30 May 12:30 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 6
HOMEBREW_SORBET_RUNTIME: set
all_proxy: socks5://127.0.0.1:7890
http_proxy: http://127.0.0.1:7890
https_proxy: http://127.0.0.1:7890
Homebrew Ruby: 3.3.1 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.1/bin/ruby
CPU: hexa-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.21.0 => /usr/local/bin/git
Curl: 8.6.0 => /usr/bin/curl
macOS: 14.5-x86_64
CLT: 15.3.0.0.1.1708646388
Xcode: 15.4
```

### Output of `brew doctor`

[brew_doctor_output.txt](https://github.com/Homebrew/discussions/files/15500303/brew_doctor_output.txt)

### Description of issue

How to update casks in taps maintained by myself, via `brew upgrade`

```
# brew info picguard

==> picguard: 1.0.0+240
https://kjxbyz.com/picguard
Installed
/usr/local/Caskroom/picguard/1.0.0+240 (122B)
From: https://github.com/picguard/homebrew-picguard/blob/HEAD/Casks/picguard.rb
==> Name
PicGuard
==> Description
Your pictures, your signature.
==> Artifacts
PicGuard.app (App)
```

```
# brew livecheck picguard 
       
picguard: 1.0.0+240 ==> 1.0.0+250
```

```
# brew outdated --cask --debug

/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromNameLoader): loading picguard
```

```
# brew upgrade picguard

Warning: Not upgrading picguard, the latest version is already installed
```

**Comment by gromgit** [comment]
`brew bump-cask-pr` is your friend, probably something like `brew bump-cask-pr --version=1.0.0+250 picguard/picguard/picguard`.

Or you could just edit `picguard.rb` yourself.

**Reply by kjxbyz** [reply]
1、Do I just need to modify the version in cask? And after the modification, I need to submit a PR to [homebrew-cask](https://github.com/Homebrew/homebrew-cask)

2、How do I format my code in development environment?

![20240531-085545](https://github.com/Homebrew/discussions/assets/47768002/1c762007-add0-47e9-ad82-261cb6892539)

**Reply by gromgit** [reply]
> Do I just need to modify the version in cask?

If nothing else (except the version#) changes about the installation process, `brew bump-cask-pr` should be all you need. The messages you're seeing are because your cask definition doesn't meet Homebrew's standard for "cleanliness".

I don't deal with casks myself, but the `[Corrected]` lines indicate that `bump-cask-pr` also ran `brew audit --strict --fix` on your cask definition, which automatically fixed some stuff, but just warned about other things that it cannot safely fix on your behalf.

I recommend you run `brew audit --strict --fix` again, fix _all_ the warnings/errors, then run `brew audit --strict --fix` again until you get no more messages. Then, if the version# in your cask is still the old one, run your `brew bump-cask-pr` again to bump the version# without any more warnings/errors.

> And after the modification, I need to submit a PR to [homebrew-cask](https://github.com/Homebrew/homebrew-cask)

Since it doesn't already exist in `homebrew/cask`, you'd be submitting a new cask, so first make sure it meets the [minimum criteria](https://docs.brew.sh/Acceptable-Casks) to avoid an automatic rejection, then follow the process detailed [here](https://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request). Failure to fix everything `brew audit` tells you about also often leads to eventual rejection.

> 2、How do I format my code in development environment?

I don't understand what you're asking here.

**Reply by kjxbyz** [reply]
1. I think it is unreasonable. I maintain homebrew tap myself, so the version update of casks under tap should be the responsibility of tap repository. What I mean is that I should update the version number of casks in tap repository, and there is no need to submit PR to homebrew-cask repository.

2. If the version number of casks in tap repository needs to be manually updated every time an update is made, what is the role of homebrew livecheck? I thought that when running brew upgrade, brew livecheck will be run first to check whether there is a version update, and if there is a new version, the update operation will be performed.

**Reply by SMillerDev** [reply]
> If the version number of casks in tap repository needs to be manually updated every time an update is made, what is the role of homebrew livecheck? I thought that when running brew upgrade, brew livecheck will be run first to check whether there is a version update, and if there is a new version, the update operation will be performed.

That sounds like you mean `brew bump --open-pr <cask>`

**Reply by kjxbyz** [reply]
Why do I have to submit a PR?

I mean, since I maintain brew taps and casks myself, why can't I submit new commits directly to the repository? For example, the current version of cask is 1.0.0+0, and the user installs 1.0.0+0. A few days later, I submit a commit to update the version to 1.0.0+1. Then when the user runs brew upgrade, the update operation should be performed.

**Reply by gromgit** [reply]
> 1. I think it is unreasonable. I maintain homebrew tap myself, so the version update of casks under tap should be the responsibility of tap repository. What I mean is that I should update the version number of casks in tap repository, and there is no need to submit PR to homebrew-cask repository.

Ah OK, then yes, there's no need to submit a `homebrew/cask` PR, and since it's your own tap:
```
brew bump-cask-pr --version=<new_version> --write-only <formula>
```
is more what you want. `brew bump-cask-pr --help` to see which options do what. (

> 2. If the version number of casks in tap repository needs to be manually updated every time an update is made, what is the role of homebrew livecheck? I thought that when running brew upgrade, brew livecheck will be run first to check whether there is a version update, and if there is a new version, the update operation will be performed.

As the name suggests, `brew livecheck` only *_checks_* for a new version of the source/binary release underlying the formula/cask. It does *_NOT_* blindly update your formula/cask, because:
1. people get upset when a command does more than what it _literally_ says it does
2. new versions may also involve changes in installation procedures, which an automated tool cannot magically fathom

And no, `brew update` is automatically run before `brew upgrade`, if it's been a while since your last `brew upgrade` and you didn't tell it not to auto-run an update. Both `update` and `upgrade` are normal user commands, while `livecheck` is a _developer_ command that's not part of normal user workflow. `livecheck` is really for Homebrew maintainers and other people like myself who maintain taps for third-party software (i.e. stuff we didn't write ourselves) to quickly discover what's been updated, including any `resource`s which the main formula/cask use.
